flaml/integrations/oai/completion.py

```diff
@@ -0,0 +1,441 @@
+        "text-davinci-002": 0.02,
+        "text-davinci-003": 0.02,
+    }
+
+    default_search_space = {
+        "model": tune.choice(list(price1K.keys())),
+        "temperature_or_top_p": tune.choice(
+            [
+                {"temperature": tune.uniform(0, 1)},
+                {"top_p": tune.uniform(0, 1)},
+            ]
+        ),
+        "max_tokens": tune.lograndint(100, 1000),
+        "n": tune.randint(1, 100),
+        "prompt": "{prompt}",
+    }
+
+    seed = 41
+    # retry after this many seconds
+    retry_time = 10
+    # fail a request after hitting RateLimitError for this many seconds
+    retry_timeout = 60
+
+    @classmethod
+    def set_cache(cls, seed=41, cache_path=".cache"):
+        """Set cache path.
+
+        Args:
+            seed (int, Optional): The integer identifier for the pseudo seed.
+                Results corresponding to different seeds will be cached in different places.
+            cache_path (str, Optional): The root path for the cache.
+                The complete cache path will be {cache_path}/{seed}.
+        """
+        cls.seed = seed
+        cls.cache_path = f"{cache_path}/{seed}"
+
+    @classmethod
+    def _get_response(cls, config: dict, eval_only=False):
+        """Get the response from the openai api call.
+
+        Try cache first. If not found, call the openai api. If the api call fails, retry after retry_time.
+        """
+        key = get_key(config)
+        response = cls._cache.get(key, None)
+        if response is not None and (response != -1 or not eval_only):
+            # print("using cached response")
+            return response
+        retry = 0
+        while eval_only or retry * cls.retry_time < cls.retry_timeout:
+            try:
+                response = openai.Completion.create(**config)
+                cls._cache.set(key, response)
+                return response
+            except (
+                ServiceUnavailableError,
+                APIError,
+                InvalidRequestError,
+                APIConnectionError,
+            ):
+                logger.info(f"retrying in {cls.retry_time} seconds...", exc_info=1)
+                sleep(cls.retry_time)
+            except RateLimitError:
+                logger.info(f"retrying in {cls.retry_time} seconds...", exc_info=1)
+                retry += 1
+        logger.warning(
+            f"Failed to get response from openai api due to getting RateLimitError for {cls.retry_timeout} seconds."
+        )
+        response = -1
+        cls._cache.set(key, response)
+        return response
+
+    @classmethod
+    def _get_max_safe_n(cls, key, max_tokens):
+        # find the max value in max_safe_n_per_max_tokens whose key is equal or larger than max_tokens
+        return max(
+            (
+                value
+                for key, value in cls._max_safe_n_per_max_tokens.get(key, {}).items()
+                if key >= max_tokens
```

Comment:
is this correct? key is both the input argument and also a local variable

Reply:
revised.